Modules/_localemodule.c

```diff
@@ -602,14 +653,24 @@ _locale_nl_langinfo_impl(PyObject *module, int item)
     /* Check whether this is a supported constant. GNU libc sometimes
        returns numeric values in the char* return value, which would
        crash PyUnicode_FromString.  */
-    for (i = 0; langinfo_constants[i].name; i++)
+    for (i = 0; langinfo_constants[i].name; i++) {
         if (langinfo_constants[i].value == item) {
             /* Check NULL as a workaround for GNU libc's returning NULL
                instead of an empty string for nl_langinfo(ERA).  */
             const char *result = nl_langinfo(item);
             result = result != NULL ? result : "";
-            return PyUnicode_DecodeLocale(result, NULL);
+            char *oldloc = NULL;
+            if (langinfo_constants[i].category != LC_CTYPE
+                && !is_all_ascii(result)
+                && change_locale(langinfo_constants[i].category, &oldloc) < 0)
+            {
+                return NULL;
+            }
```

Comment:
You don't call nl_langinfo() twice anymore? Why changing the code?

Reply:
Because I now think that it was not needed. The result can only be changed if `setlocale()` is called with the same category or LC_ALL. And, even if it is not explicitly stated, I think that calls `setlocale(..., NULL)` are not counted.

On other hand,I want to avoid duplication of the code for conversion of the result because of #124974.

Reply:
`locale.localeconv()` calls `localeconv()` only once, whereas it changes the LC_CTYPE locale while reading its output. Let's do the same for `nl_langinfo()`. If the assumption is wrong, we can adjust the code later.